hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/StorageBasedLockProvider.java

```diff
@@ -0,0 +1,581 @@
+import static org.apache.hudi.common.table.HoodieTableMetaClient.LOCKS_FOLDER_NAME;
+
+/**
+ * A distributed filesystem storage based lock provider. This {@link LockProvider} implementation
+ * leverages conditional writes to ensure transactional consistency for multi-writer scenarios.
+ * The underlying storage client interface {@link StorageLock} is pluggable so it can be implemented for any
+ * filesystem which supports conditional writes.
+ */
+@ThreadSafe
+public class StorageBasedLockProvider implements LockProvider<StorageLockFile> {
+
+  public static final String DEFAULT_TABLE_LOCK_FILE_NAME = "table_lock";
+  // How long to wait before retrying lock acquisition in blocking calls.
+  private static final long DEFAULT_LOCK_ACQUISITION_BUFFER_MS = 1000;
+  // Maximum expected clock drift between two nodes.
+  // This is similar idea as SkewAdjustingTimeGenerator.
+  // In reality, within a single cloud provider all nodes share the same ntp
+  // server
+  // therefore we do not expect drift more than a few ms.
+  // However, since our lock leases are pretty long, we can use a high buffer.
+  private static final long CLOCK_DRIFT_BUFFER_MS = 500;
+
+  // When we retry lock upserts, do so 5 times
+  private static final long LOCK_UPSERT_RETRY_COUNT = 5;
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(StorageBasedLockProvider.class);
+
+  Logger logger;
+
+  // The lock service implementation which interacts with storage
+  private final StorageLock lockService;
+
+  private final long heartbeatInterval;
+  private final long lockValidity;
+  private final String ownerId;
+  private final String lockFilePath;
+  private final String bucketName;
+  private final HeartbeatManager heartbeatManager;
+  private transient Thread shutdownThread = null;
+
+  @GuardedBy("this")
+  private StorageLockFile currentLockObj = null;
+  @GuardedBy("this")
+  private boolean isClosed = false;
+
+  private synchronized void setLock(StorageLockFile lockObj) {
+    if (lockObj != null && !Objects.equals(lockObj.getOwner(), this.ownerId)) {
+      throw new HoodieLockException("Owners do not match. Current lock owner: " + this.ownerId + " lock path: "
+          + this.lockFilePath + " owner: " + lockObj.getOwner());
+    }
+    this.currentLockObj = lockObj;
+  }
+
+  /**
+   * Default constructor for StorageBasedLockProvider, required by LockManager
+   * to instantiate it using reflection.
+   * 
+   * @param lockConfiguration The lock configuration, should be transformable into
+   *                          StorageBasedLockConfig
+   * @param conf              Storage config, ignored.
+   */
+  public StorageBasedLockProvider(final LockConfiguration lockConfiguration, final StorageConfiguration<?> conf) {
+    StorageBasedLockConfig config = new StorageBasedLockConfig.Builder()
+        .fromProperties(lockConfiguration.getConfig()).build();
+    heartbeatInterval = config.getHeartbeatPoll();
```

Comment:
again, lets try to align the variable naming and getters. causes confusion if we are referring to two diff things